pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -398,6 +411,166 @@ func (c *Controller) setupIndex() *AmbientIndex {
 			}
 		},
 	}
+
+	idx.servicesMap = make(map[types.NamespacedName]*model.Service)
+	idx.handleServiceEntry = func(svc *model.Service, event model.Event) {
+		idx.mu.Lock()
+		defer idx.mu.Unlock()
+
+		if svc.Attributes.ServiceEntry == nil {
+			// event for e.g. kube svc; ignore
+			return
+		}
+
+		// We will accrue updates as we update our internal state
+		updates := sets.New[model.ConfigKey]()
+
+		if event != model.EventAdd {
+			c.cleanupOldServiceEntryVips(svc, updates)
+		}
+
+		serviceEntryNamespacedName := types.NamespacedName{
+			Name:      svc.Attributes.ServiceEntryName,
+			Namespace: svc.Attributes.ServiceEntryNamespace,
+		}
+
+		// Update indexes
+		if event == model.EventDelete {
+			// servicesMap is used when cleaning up old VIPs (e.g. `ServiceEntry.endpoints`) so we must
+			// delete this after we clean up VIPs
+			delete(idx.servicesMap, serviceEntryNamespacedName)
+		} else {
+			// servicesMap is used when constructing workloads so it must be up to date
+			c.ambientIndex.servicesMap[serviceEntryNamespacedName] = svc
+		}
+
+		pods := c.podsClient.List(metav1.NamespaceAll, klabels.Everything())
+		wls := make(map[string]*model.WorkloadInfo, len(pods))
+		for _, pod := range pods {
+			newWl := c.extractWorkload(pod)
+			if newWl != nil {
+				// Update the pod, since it now has new VIP info
+				networkAddrs := networkAddressFromWorkload(newWl)
+				for _, networkAddr := range networkAddrs {
+					c.ambientIndex.byPod[networkAddr] = newWl
+				}
+				c.ambientIndex.byUID[c.generatePodUID(pod)] = newWl
+				updates[model.ConfigKey{Kind: kind.Address, Name: newWl.ResourceName()}] = struct{}{}
```

Comment:
`Insert` should be used here

Reply:
and L480   L492

Reply:
addressed in https://github.com/istio/istio/pull/45621/commits/607d0b1d6a6961d93a2e6a45b5b33f86cdd99698